apex/normalization/csrc/layer_norm_cuda_kernel.cu

```diff
@@ -0,0 +1,975 @@
+        k_grad_input[l] = static_cast<T>(f_grad_input);
+      }
+    } else {
+      for (int l = thrx;  l < n2;  l+=numx) {
+        const U c_h = static_cast<U>(k_input[l]);
+        const U c_loss = static_cast<U>(k_dout[l]);
+        U f_grad_input = fH * c_loss;
+        f_grad_input -= sum_loss1;
+        f_grad_input -= (c_h - c_mean) * c_invvar * sum_loss2;
+        f_grad_input *= term1;
+        k_grad_input[l] = static_cast<T>(f_grad_input);
+      }
+    }
+  }
+}
+
+template<typename T> 
+void HostApplyLayerNorm(
+    T* output,
+    at::Tensor* mean,
+    at::Tensor* invvar,
+    const T* input,
+    int n1,
+    int n2,
+    double epsilon,
+    const T* gamma,
+    const T* beta
+    )
+{
+    auto stream = at::cuda::getCurrentCUDAStream().stream();
+    const dim3 threads(32,4,1);
+    const dim3 blocks(1,n1,1);
+    int nshared = 
+        threads.y > 1 ? 
+	    threads.y*sizeof(float)+(threads.y/2)*sizeof(float) : 
+	    0;
+    cuApplyLayerNorm<<<blocks, threads, nshared, stream>>>(
+		    output,
+		    mean->data<float>(),
+		    invvar->data<float>(),
+		    input,
+		    n1,n2,
+		    (float)epsilon,
+                    gamma,beta);
+}
+template<> 
+void HostApplyLayerNorm(
+    int64_t* output,
+    at::Tensor* mean,
+    at::Tensor* invvar,
+    const int64_t* input,
+    int n1,
+    int n2,
+    double epsilon,
+    const int64_t* gamma,
+    const int64_t* beta
+    )
+{
+    auto stream = at::cuda::getCurrentCUDAStream().stream();
+    const dim3 threads(32,4,1);
+    const dim3 blocks(1,n1,1);
+    int nshared = 
+        threads.y > 1 ? 
+	    threads.y*sizeof(double)+(threads.y/2)*sizeof(double) : 
+	    0;
+    cuApplyLayerNorm<<<blocks, threads, nshared, stream>>>(
+		    output,
+		    mean->data<double>(),
+		    invvar->data<double>(),
+		    input,
+		    n1,n2,
+		    (double)epsilon,
+                    gamma,beta);
+}
+template<> 
+void HostApplyLayerNorm(
+    double* output,
+    at::Tensor* mean,
+    at::Tensor* invvar,
+    const double* input,
+    int n1,
+    int n2,
+    double epsilon,
+    const double* gamma,
+    const double* beta
+    )
+{
+    auto stream = at::cuda::getCurrentCUDAStream().stream();
+    const dim3 threads(32,4,1);
+    const dim3 blocks(1,n1,1);
+    int nshared = 
+        threads.y > 1 ? 
+	    threads.y*sizeof(double)+(threads.y/2)*sizeof(double) : 
+	    0;
+    cuApplyLayerNorm<<<blocks, threads, nshared, stream>>>(
+		    output,
+		    mean->data<double>(),
+		    invvar->data<double>(),
+		    input,
+		    n1,n2,
+		    (double)epsilon,
+                    gamma,beta);
+}
+
+void cuda_layer_norm(
+    at::Tensor* output,
+    at::Tensor* mean,
+    at::Tensor* invvar,
+    at::Tensor* input,
+    int n1,
+    int n2,
+    at::IntList normalized_shape,
+    at::Tensor* gamma,
+    at::Tensor* beta,
+    double epsilon)
+{
+    AT_DISPATCH_ALL_TYPES_AND_HALF(input->type(), "layer_norm_cuda_kernel", ([&] {
+        HostApplyLayerNorm(
```

Comment:
there's no need to dispatch for integer types (normalization operation is ill-defined on integer types anyway), and to avoid specializations, you can use accscalar_t inside dispatch macro (`using accscalar_t = acc_type<scalar_t, true>;`), it defines accscalar_t to float when scalar_t is at::Half, and to scalar_t otherwise (when scalar_t is float or double). Same applies to backward.

Reply:
I like that, it will simplify the code a little bit.

Reply:
You need `#include "ATen/AccumulateType.h"` for this